fix #41546, make `using` thread-safe

JeffBezanson opened the pull request:
fix #41546

I'm not sure if `identify_package` needs locking as well.

JeffBezanson committed:
fix #41546, make `using` thread-safe

Sacha0 reviewed (commented):
Thanks so much Jeff! This is exciting! I will pull this patch onto our testing branch and report back :). Thanks again!

vtjnash commented on base/toml_parser.jl:
```diff
@@ -104,7 +104,7 @@ function Parser(str::String; filepath=nothing)
             root,
             filepath,
-            isdefined(Base, :loaded_modules) ? get(Base.loaded_modules, DATES_PKGID, nothing) : nothing,
+            isdefined(Base, :loaded_modules) ? Base.maybe_root_module(DATES_PKGID) : nothing,
```
```suggestion
            isdefined(Base, :maybe_root_module) ? Base.maybe_root_module(DATES_PKGID) : nothing,
```

JeffBezanson committed:
Update base/toml_parser.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Sacha0 commented:
With this patch our test suite silently times out after launching tests, which may or may not be progress? 😄

vtjnash commented on base/loading.jl:
```diff
@@ -1020,13 +1024,14 @@ function require(uuidkey::PkgId)
         end
```
seems suspect that we want to be holding a lock for this large of a chunk of (user) code, including the `include` and `package_callbacks` here, but we do need it for all of the shared state, which is quite a bit.

vtjnash committed:
squash! fix #41546, make `using` thread-safe

use more precision when handling loading lock, merge with TOML lock
(since we typically are needing both, sometimes in unpredictable
orders), and unlock before call most user code

timholy commented:
`git bisect` identified this as at least part of what broke Revise on nightly. Here's a replication that doesn't take too long to run:

```sh
julia -e 'import Pkg; Pkg.test("Revise";coverage=false, test_args=["New files & Requires.jl"])'
```

mkitti commented:
I believe this commit was responsible for stalling HDF5.jl's nightly testing for the past two months:
https://github.com/JuliaIO/HDF5.jl/runs/4018361152?check_suite_focus=true

The symptom was that the CI would stall and eventually timeout after 6 months. The intermediate solution was to add a 20 minute timeout to CI:
https://github.com/JuliaIO/HDF5.jl/issues/877

<details>
<summary>

Upon debugging, I found the issue came when trying to log external compression codecs when [testing filters](https://github.com/JuliaIO/HDF5.jl/pull/880/files#diff-3b9314a6f9f2d7eec1d0ef69fa76cfabafdbe6d0df923768f9ec32f27a249c63R42-R43)

</summary>

```julia
[ Info: libhdf5 v1.12.0
┌ Debug: plain
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\runtests.jl:14
┌ Debug: compound
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\runtests.jl:16
┌ Debug: custom
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\runtests.jl:18
┌ Debug: reference
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\runtests.jl:20
┌ Debug: dataspace
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\runtests.jl:22
┌ Debug: hyperslab
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\runtests.jl:24
┌ Debug: readremote
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\runtests.jl:26
┌ Debug: extend_test
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\runtests.jl:28
┌ Debug: gc
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\runtests.jl:30
┌ Debug: external
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\runtests.jl:32
┌ Debug: external tests did not delete
│   fn2 = "C:\\cygwin64\\tmp\\jl_TuY7eOkBtL"
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\external.jl:57
┌ Debug: swmr
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\runtests.jl:34
┌ Debug: mmap
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\runtests.jl:36
┌ Debug: properties
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\runtests.jl:38
┌ Debug: table
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\runtests.jl:40
┌ Debug: filter
└ @ Main C:\Users\kittisopikulm\.julia\dev\HDF5\test\runtests.jl:42
# Then it deadlocks here.
```
</details>

After adding some debug messages to trace package loading, I found that a second process was trying to load one of the compression codecs via the H5Zbzip2.jl subdirectory package, which loads CodecBzip2.jl and Bzip2_jll. The second process came from the [swmr.jl tests](https://github.com/JuliaIO/HDF5.jl/pull/880/files#diff-d6f848692d47d2ab2f25684c10ad8e69ea57023b1688e4c0aa13f0b2d410282aR8) which used `Distributed.addproc(1)` to add a process. [Removing the second process resolved the issue](https://github.com/JuliaIO/HDF5.jl/pull/880/files#diff-d6f848692d47d2ab2f25684c10ad8e69ea57023b1688e4c0aa13f0b2d410282aR121) solved the stall in the tests.

 Commenting out `rmprocs` is sufficient to cause the stall.

I have created a branch / draft pull request on HDF5.jl for the purpose of investigating this issue:
https://github.com/JuliaIO/HDF5.jl/pull/881

Sacha0 commented:
Possibly related to https://github.com/JuliaLang/julia/pull/43339#issuecomment-990222989
